File: sources/csharp/03_func_sum/main.cs
// task 03 func_sum — expected output: 100000000
// build: coreclr: dotnet build -c Release | nativeaot: dotnet publish -c Release -p:PublishAot=true | mono: mcs -optimize+ main.cs    run: coreclr: dotnet run -c Release (or bin/Release/net8.0/main.exe) | nativeaot: bin/Release/net8.0/publish/main.exe | mono: mono main.exe
using System;
using System.Runtime.CompilerServices;

class Program
{
    // NoInlining keeps the call real; without it the JIT would fold the whole loop away.
    [MethodImpl(MethodImplOptions.NoInlining)]
    static long AddOne(long n)
    {
        return n + 1;
    }

    static void Main()
    {
        long value = 0;
        for (int i = 0; i < 100000000; i++)
        {
            value = AddOne(value);
        }
        Console.WriteLine(value);
    }
}

File: sources/fsharp/03_func_sum/main.fs
// task 03 func_sum — expected output: 100000000
// build: dotnet build -c Release    run: dotnet run

open System.Runtime.CompilerServices

// Not inlinable, so the hundred million calls really happen.
[<MethodImpl(MethodImplOptions.NoInlining)>]
let addOne (n: int64) : int64 = n + 1L

let mutable value = 0L
let mutable i = 0
while i < 100000000 do
    value <- addOne value
    i <- i + 1

printfn "%d" value
